File: Data Structures/Tree-based-DSAs/Binary-tree/valid_binary_search_tree.py
"""
    validate a binary search tree:
        1. duplicate values invalidate the binary tree
    

"""

"""
    space and time analysis:
        space: O(n)
        time: O(n)

"""
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isValidBST(self, root: TreeNode):
    return validate_bst(root)

def validate_bst(root : TreeNode, maximum = float('inf'), minimum = float('-inf')):
    if not root:
        return True
    if root.val <= minimum  or root.val >= maximum:
        return False
    return validate_bst(root.left, root.val, minimum) and \
        validate_bst(root.right, maximum, root.val) 

File: Data Structures/Tree-based-DSAs/Binary-tree/valid-binary-search-tree.cpp
#include "bits/stdc++.h"
 /* Definition for a binary tree node.*/
 struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };
 

using namespace std;

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (!root)return true;

       return dfs(root, numeric_limits<long long int>::min(), numeric_limits<long long int>::max());
    }

    bool dfs(TreeNode * root, long long int min, long long max){
        /* return if we are outside of our boundaries */
        if (!root) return true;

        if(root->val <= min || root->val >= max) return false;
        return dfs(root->left, min, root->val) && dfs(root->right, root->val, max);
    }

};

int main(){
    TreeNode node(2147483647);
    Solution instance;
    cout << instance.isValidBST(&node) <<endl;
}